src/index.ts

```diff
@@ -306,51 +316,65 @@ QuickSQLite.transactionAsync = (
   };
 
   const commit = () => {
+    if (isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute commit on finalized transaction: ${dbName}`
+      );
+    }
     const result = QuickSQLite.execute(dbName, 'COMMIT');
     isFinalized = true;
     return result;
   };
 
   const rollback = () => {
+    if (isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute rollback on finalized transaction: ${dbName}`
+      );
+    }
     const result = QuickSQLite.execute(dbName, 'ROLLBACK');
     isFinalized = true;
     return result;
   };
 
-  const tx: PendingTransaction = {
-    start: async () => {
-      try {
-        QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
-        await callback({
-          commit,
-          execute,
-          executeAsync,
-          rollback,
-        });
-
-        if (!isFinalized) {
-          commit();
-        }
-      } catch (e: any) {
-        if (!isFinalized) {
-          rollback();
-        }
+  return await new Promise((resolve, reject) => {
+    const tx: PendingTransaction = {
+      start: async () => {
+        try {
+          QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
+          await callback({
+            commit,
+            execute,
+            executeAsync,
+            rollback,
+          });
+
+          if (!isFinalized) {
+            commit();
+          }
 
-        // Do not throw an error, because the transaction is executed with a setImmediate call
-        // This errors are uncatchable
-        // https://stackoverflow.com/questions/51081892/nodejs-asynchronous-exceptions-are-uncatchable
+          resolve();
+        } catch (e) {
+          if (!isFinalized) {
+            try {
+              rollback();
+            } catch (rollbackError) {
+              reject(rollbackError);
+            }
+          }
 
-        // throw e;
-      } finally {
-        locks[dbName].inProgress = false;
-        isFinalized = false;
-        startNextTransaction(dbName);
-      }
-    },
-  };
+          reject(e);
+        } finally {
+          locks[dbName].inProgress = false;
+          isFinalized = false;
+          startNextTransaction(dbName);
+        }
+      },
+    };
 
-  locks[dbName].queue.push(tx);
-  startNextTransaction(dbName);
+    locks[dbName].queue.push(tx);
```

Comment:
Are you sure the order of the transactions does not get affected by wrapping the push inside of the promise?

Reply:
It queues them up correctly.
When constructing a promise with a callback, the callback is called as part of the constructor.

See this [sandbox](https://codesandbox.io/s/queued-promise-settlement-timing-tmbv9f?file=/src/index.js) to show the sequence.